src/af_tsa3.c

```diff
@@ -1017,238 +44,50 @@ static int tsa_swap(struct sk_buff *skb, struct genl_info *info)
-		}
-	} else {
-		net = get_net(current->nsproxy->net_ns);
-	}
-
-	err = __sock_create(net, domain, type, protocol, &underlyingsock, 0);
-	if (err) {
-		NL_SET_ERR_MSG_MOD(info->extack, "could not create underlying socket");
-		goto out_put_net;
-	}
-
-	reply = nlmsg_new(NLMSG_DEFAULT_SIZE, GFP_KERNEL);
-	if (!reply) {
-		NL_SET_ERR_MSG_MOD(info->extack, "could not allocate memory for reply");
-		err = -ENOMEM;
-		goto out_put_sock;
-	}
-
-	if (!try_module_get(THIS_MODULE)) {
-		NL_SET_ERR_MSG_MOD(info->extack, "could not fetch module failed");
-		err = -EBUSY;
-		goto out_put_msg;
-	}
-
-	/* TODO: Figure this out
-	err = security_socket_create(domain, type, protocol, 0);
-	if (err) {
-		NL_SET_ERR_MSG_MOD(info->extack,
-				   "security_socket_create failed");
-		goto out_put_module;
-	}
-	*/
-
-	sock = sock_alloc();
-	if (!sock) {
-		NL_SET_ERR_MSG_MOD(info->extack, "socket: no more sockets");
-		err = -ENFILE;
-		goto out_put_module;
-	}
-	pr_debug("TSA creation of top level socket: %px\n", sock);
-
-	sock->type = type;
-
-	WARN_ON(!sock);
-	WARN_ON(!underlyingsock);
-	newfile = sock_alloc_file(sock, flags, NULL);
-	if (IS_ERR(newfile)) {
-		NL_SET_ERR_MSG_MOD(info->extack, "sock_alloc_file failed");
-		goto out_put_tsa_socket;
-	}
-
-	tpow = kmem_cache_zalloc(tsa_tpow_cachep, GFP_KERNEL);
-	if (!tpow) {
-		NL_SET_ERR_MSG_MOD(info->extack, "could not allocate tsa socket operations wrapper");
-		err = -ENOMEM;
-		goto out_put_tsa_socket;
-	}
-
-	err = init_srcu_struct(&tpow->srcu);
-	if (err) {
-		NL_SET_ERR_MSG_MOD(info->extack, "could not allocate srcu");
-		kmem_cache_free(tsa_tpow_cachep, tpow);
-		goto out_put_tsa_socket;
-	}
-	INIT_WORK(&tpow->work_free, __tsa_free_tpow_cb);
-	kref_init(&tpow->kref);
-
-	tsa_make_wrapper(tpow, domain);
-	sock->ops = &tpow->real_ops;
-	WARN_ON(!sock->ops);
-
-	err = tsa_make_socket(sock, underlyingsock);
-	if (err) {
-		NL_SET_ERR_MSG_MOD(info->extack, "Could not create socket");
-		goto out_put_file;
-	}
-
-	/* TODO: Figure this out
-	err = security_socket_post_create(sock, domain, type, protocol, 0);
-	if (err) {
-		NL_SET_ERR_MSG_MOD(info->extack,
-				   "security_socket_post_create failed");
-		goto out_put_socket;
-	}
-	*/
-
-	fd = get_unused_fd_flags(flags);
-	if (fd < 0) {
-		NL_SET_ERR_MSG_MOD(info->extack, "get_unused_fd_flags failed");
-		err = fd;
-		goto out_put_file;
+	// Ensure socket is in an unconnected state and is
+	// not bound to a device to ensure switching network
+	// namespaces does not result in undesired behavior.
+	sk = sock->sk;
+	lock_sock(sk);
+	if (sk->sk_state != TCP_CLOSE) {
+		NL_SET_ERR_MSG_MOD(info->extack, "Socket is not in an unconnected state");
+		goto out_put_release;
 	}
 
-	err = tsa_create_reply(fd, reply, info);
-	if (!err) {
-		fd_install(fd, newfile);
-		err = 0;
-		goto out_put_net;
+	if (sk->sk_bound_dev_if != 0) {
+		NL_SET_ERR_MSG_MOD(info->extack, "Socket is bound to a device");
+		goto out_put_release;
 	}
+	mine = sock_net(sk);
```

Comment:
Can we change the variable names to `my_ns` and the other to `other_ns`?

Reply:
fixed